packages/mongo/oplog_tailing.js

```diff
@@ -121,15 +121,51 @@ _.extend(OplogHandle.prototype, {
   // currently visible.
   // XXX become convinced that this is actually safe even if oplogConnection
   // is some kind of pool
-  waitUntilCaughtUp: function () {
+  _oplogCheckRunning: false,
+  _oplogCheckCallbacks: [],
+  waitUntilCaughtUp: function() {
     var self = this;
-    if (self._stopped)
-      throw new Error("Called waitUntilCaughtUp on stopped handle!");
-
     // Calling waitUntilCaughtUp requries us to wait for the oplog connection to
     // be ready.
     self._readyFuture.wait();
 
+    if (self._stopped) {
+      throw new Error("Called waitUntilCaughtUp on stopped handle!");
+    }
+    
+    var f = new Future();
+    self._oplogCheckCallbacks.push(f.return.bind(f));
+    self._tryFindingOplogLastEntry();
+    f.wait();
+  },
+  _tryFindingOplogLastEntry: function () {
+    var self = this;
+    
+    if(self._oplogCheckRunning) {
+      return;
+    }
+
+    var callbacks = self._oplogCheckCallbacks;
+    if(callbacks.length === 0) {
+      return;
+    }
+
+    function startProcess() {
```

Comment:
Not sure how this function helps clarity — it moves the read of `_oplogCheckCalbacks` and its reset farther away from each other in the source.